pkg/operator/ceph/cluster/controller.go

```diff
@@ -256,6 +257,19 @@ func (r *ReconcileCephCluster) reconcile(request reconcile.Request) (reconcile.R
 		doCleanup := true
 		logger.Infof("deleting ceph cluster %q", cephCluster.Name)
 
+		// TODO: more work to check if data exists in the cluster
+		subresources := subresource.CephClusterRegistry.List()
```

Comment:
Why do we need an interface for subresources? It seems pretty simple to just directly count the number of CephFilesystem, CephObjectStore, or CephBlockStore CRs. IMO avoiding interfaces is usually a better design, although in some cases they certainly are needed.

Reply:
Interfaces allow us to delegate the work to the subresources themselves. IMO, that's where the logic belongs. If we add another subresource type, nothing needs added to this code, and the code is less tightly coupled. It also helps us reuse code and keep a similar and predictable code structure for all resources. It's going to get really messy if each controller has a different way of checking its subresources.

It's pretty simple for CephCluster-depends-on-CephFilesystem because it's just a matter of checking if FS resources exist. But it's harder for CephNFS which can be dependent on a CephFilesystem, a CephBlockPool, OR a CephObjectStore. And CephNFS doesn't reference the FS/block/object resource itself but rather a pool created by the resource.

It's also harder for OBCs which can be in ANY namespace and which also don't reference a CephObjectStore but instead a StorageClass which in turn references the CephObjectStore.

Reply:
In Rook, the parent CRD knows who all its dependents are. This isn't an openly extensible system. 
- CephCluster --> CephFilesystem, CephBlockPool, CephObjectStore
- CephObjectStore --> CephObjectZone
- CephObjectZone --> CephObjectZoneGroup
- CephObjectZoneGroup --> CephObjectRealm
- CephFilesystem --> CephNFS

We don't need to replace anything at runtime (e.g. mocking) with different types, which is really where interfaces are most useful.  

For the purposes of finalizers, I'm not sure it's a good idea to reach outside to other namespaces and enumerate all the OBCs, RBD volumes, and CephFS volumes. Instead, querying Ceph to see if those buckets and volumes exist seems sufficient.

Reply:
I disagree that interfaces are only useful for mocking or allowing extensibility. Interfaces allow for code that is more decoupled and where systems can be more independent of changes to other parts of the code, which is what I try to do here. I will keep thinking about this, but I still don't think it's wrong to delegate logic for checking if resources exist to the controller that manages the resources. I think that is good design.

As a note, I think you have the dependencies on objects/zones backwards though. The CephObjectStore references (depends on) CephObjectZone, not the other way around.

Also, CephNFS references any pool, which could be CephFilesystem, CephBlockPool, CephObjectStore, or a manually-created pool. This means the CephCluster also should check to see if there are CephNFSes that exist (which could reference manual pools). The CephBlockPool and CephObjectStore also should check.

Reply:
For interfaces, one of my concerns is that it makes the code harder to navigate. Go doesn't make implementers of interfaces easy to discover. If there isn't a real **need** for an interface, the downsides outweigh the benefits. Perhaps I am soured to interfaces because a long time ago (>10 yrs) I worked with someone who loved pure design principles and used interfaces everywhere. The code was beautiful to him, but very difficult for others to maintain.

Decoupling isn't a reason that I see necessary in Rook since it's a closed system. Decoupling is useful for extensibility, but it also has a cost of increased code complexity.

About the only place Rook currently defines interfaces (excluding generated code) is to enable test code:
- `CephManifests`: Enabled upgrade testing to use a common set of functions to install with different versions
- `Executor`: Enables mocking of the ceph and other commands

I feel no remorse that we don't defined more interfaces in Rook. :)
 
> As a note, I think you have the dependencies on objects/zones backwards though. The CephObjectStore references (depends on) CephObjectZone, not the other way around.

Right, good catch

Reply:
I think I'm confused because all I'm doing is dependency inversion. Also, the Go tooling with VSCode handles navigating interfaces and implementations much better today than it did even a couple years ago. I don't see this as a hugely abstract or confusing thing. In my mind, it just enforces that the function call looks the same for all subresources and makes the code look predictable. I think it will also allow for some more code reuse than if an interface didn't exist by the time I'm done.

Reply:
I'd be curious to see an example of the difference in code between using an interface vs not using an interface before we go too far with this pattern.

For example, without using an interface, the implementation would look something like this in the CephCluster finalizer (error handling is obviously missing), right?
```go
filesystems, err := clusterdCtx.RookClientset.CephV1().CephFilesystems(ns).List(ctx, metav1.ListOptions{})
if len(filesystems.Items) > 0 {
  // prevent cephcluster CR removal
}
objectStores, err := clusterdCtx.RookClientset.CephV1().CephObjectStores(ns).List(ctx, metav1.ListOptions{})
if len(objectStores.Items) > 0 {
  // prevent cephcluster CR removal
}
pools, err := clusterdCtx.RookClientset.CephV1().CephBlockPools(ns).List(ctx, metav1.ListOptions{})
if len(pools.Items) > 0 {
  // prevent cephcluster CR removal
}
```

Reply:
You are focused only on the dependency relationship where CephCluster is referenced by those 3 resources, but that is far from the case we have in reality.

For NFS it is a little bit more complicated. A pool created by a CephFilesystem, CephBlockPool, or CephObjectStore can be referenced by an NFS. So from each of those 3 controllers, we would have to...
1. Get the NFSes in the namespace
2. Read the `pool` on the NFS to determine if it is a pool provided by this resource which can be any of: CephFilesystem, CephObjectStore, or CephBlockPool

AND, to support NFSes that reference manually-created pools we also have to add NFS listing to the your example that is checked by the CephCluster controller.

For the OBC, it is much, much more complicated even though it only refers to CephObjectStores because it also goes through a StorageClass. In the CephObjectStore controller, we would have to...
1. Get the OBCs cluster-wide
2. Read the OBC parameter that tells what StorageClass it is using
3. Get the StorageClass
4. Read the appropriate StoragClass parameters to determine if it references a given CephObjectStore
(The information above is already coded into the OBC controller, which should be the thing to know/understand how it references things rather than relegating that into the parent controller)

This all starts to explode if we treat resources dependencies as special snowflake cases when we actually have many dependency relationships in Rook. We have 12 controllers in Rook, and they have different dependency relationships. 

I think it is most appropriate to delegate the knowledge about reference to the dependents since the information about how they can introspect is already coded there. Furthermore, I think we will start to have "dependency hell" if we also start doing things like importing the NFS code into CephCluster, CephFilesystem, CephObjectStore, and CephBlockPool. Not to mention that the replication controllers have flipped dependencies: The CephObjectStore references/depends on the ObjectZone, so ObjectZone also needs to import CephObjectStore and check that in its controller. Similarly, CephBlockPool references/depends on RBDMirror, so RBDMirror would need to be the one importing CephBlockPool.

The way to avoid "dependency hell" here and delegate knowledge to the dependents is to use dependency inversion.

Reply:
I could imagine implementing all of that complexity without interfaces, it's just a question of package layout to avoid circular go references. But I'm not saying an interface wouldn't be reasonable either. The benefits of the abstraction just need to be worth the added complexity.

At the end of the day, how thorough do we really need to be with finalizers? A finalizer doesn't prevent a resource from being marked for deletion, it just gives the controller an opportunity to clean up underlying resources or temporarily block deletion of the resource until other pre-reqs are met. It doesn't seem reasonable to traverse all possible resources to see if anyone is using a resource. We can have reasonable confidence that a resource is in use with some simple checks and shouldn't need to walk complex dependencies. For example, let's just check if there are any buckets in the object store instead of walking all the OBCs and analyzing their storage classes. The latter seems like overkill. IMO we should just be checking for relevant k8s resources in our own namespace or Ceph resources. Doing more than that feels like hitting a small nail with a sledgehammer.

Reply:
As a note, there aren't finalizers in this implementation. I started this PR because we all agreed to not use finalizers.

For object stores, I again get stuck thinking it is reasonable that users could wish to delete the object store without deleting all the resources kept in it first. In the COSI upstream discussions, they have talked a lot about users' ephemeral storage desires, and I don't want to force users to modify their applications to do cleanup when it isn't necessary. There should be a way for users to just be done with the storage and have it be deleted for them. I think that would mean we would need a way to force delete that we don't have currently that then would clean up the Ceph resources.

I also don't want to think about having inter-resource dependencies half-way and do our users a bit of a disservice. If we have inter-resource dependencies, the feature that users get is that their data that is in-use can be better protected from accidental deletion. I don't think it's much of a feature that we can document well if only some resources are protected some of the time. Otherwise it gets confusing to know which resources have protection and which don't as a user who doesn't know the code.

But I think we could think of a "simplest form" of this. We could check if there are any subresources existing that reference the current CephCluster to prevent CephCluster deletion but not bother with subresources between things like CephFS-NFS or CephObjectStore-OBC. And that could be a pretty simple check (except in the case of OBCs, which aren't namespaced).

Reply:
> As a note, there aren't finalizers in this implementation. I started this PR because we all agreed to not use finalizers.

Agreed, we are not changing finalizers. I'm just referring to the discussion as the implementation of the code that responds to the deletion event because we have declared a finalizer. Without the finalizers, that code isn't triggered.
 
> For object stores, I again get stuck thinking it is reasonable that users could wish to delete the object store without deleting all the resources kept in it first. In the COSI upstream discussions, they have talked a lot about users' ephemeral storage desires, and I don't want to force users to modify their applications to do cleanup when it isn't necessary. There should be a way for users to just be done with the storage and have it be deleted for them. I think that would mean we would need a way to force delete that we don't have currently that then would clean up the Ceph resources.

Actually, I wonder if we need to do anything else for object store cleanup. See [verifyObjectBucketCleanup()](https://github.com/rook/rook/blob/bb5f23323341af10924e30279474a8769684c791/pkg/operator/ceph/object/controller.go#L479) that already checks for OBs. If we're checking for OBs now I'm not sure why we need to check for OBCs. What exactly was the reported issue?

> I also don't want to think about having inter-resource dependencies half-way and do our users a bit of a disservice. If we have inter-resource dependencies, the feature that users get is that their data that is in-use can be better protected from accidental deletion. I don't think it's much of a feature that we can document well if only some resources are protected some of the time. Otherwise it gets confusing to know which resources have protection and which don't as a user who doesn't know the code.
> 
> But I think we could think of a "simplest form" of this. We could check if there are any subresources existing that reference the current CephCluster to prevent CephCluster deletion but not bother with subresources between things like CephFS-NFS or CephObjectStore-OBC. And that could be a pretty simple check (except in the case of OBCs, which aren't namespaced).

Agreed, let's keep this simple. Let's chat more...